src/schema.rs

```diff
@@ -0,0 +1,327 @@
+  None,
+  /// Represents the absence of any value.
+  Null,
+  /// Represents a binary true/false value.
+  Boolean,
+  /// Represents a numeric type.
+  Number {
+    /// Forces the number to be a multiple of another. This helps in specifying
+    /// integers if this value is `Some(1)` for example.
+    multiple_of: Option<f32>,
+    /// The minimum value the number can be.
+    minimum: Option<f64>,
+    /// The maximum value the number can be.
+    maximum: Option<f64>
+  },
+  String {
+    /// The mimimum length of characters in the string.
+    min_length: Option<u64>,
+    /// The maximum length of characters in the string.
+    max_length: Option<u64>,
+    /// A regular expression pattern to validate the string against.
+    // TODO: Use a regex crate.
+    pattern: Option<String>
+  },
+  /// Represents a set of any type.
+  Array {
+    /// A schema which all items in the array must match.
+    items: Box<Schema>
+  },
+  /// Represents any tuple of values.
+  Tuple {
+    /// Schemas which each tuple value (in the same place) must comply with.
+    items: Vec<Schema>,
+    /// Whether or not there can be more items in the tuple.
+    additional_items: bool
+  },
+  /// Represents a set of key/value pairs.
+  Object {
+    /// Schemas associated to the object properties.
+    properties: LinearMap<Key, Schema>,
+    /// Properties that are required to be in the object.
+    required: Vec<Key>,
+    /// Whether or not there may be extra properties outside of the ones
+    /// defined by the properties map.
+    additional_properties: bool
+  },
+  /// Represents a value which *must* be one of the defined values. An enum is
+  /// considered a primitive type as if it is a single value is a higher order
+  /// type, no variation is allowed.
+  Enum(Vec<Value>)
+}
+
+impl Schema {
+  /// Validates a query that a user would like to make on the database by
+  /// comparing it to the schema. Mostly checks that all properties described
+  /// in the query are accessible according to the schema.
+  pub fn validate_query(&self, query: &Query) -> Result<(), Error> {
+    static NO_PRIMITIVE_HINT: &'static str = "Try not querying specific properties of a primitive like `null` or `boolean`.";
+    lazy_static! { static ref INTEGER_RE: Regex = Regex::new(r"^\d+$").unwrap(); }
+
+    match (self, query) {
+      // No schema describes these values, its the wild west. Go crazy query.
+      // `Schema::None` does not represent the absence of value, just the
+      // absence of validation.
+      (&Schema::None, _) => Ok(()),
+      (&Schema::Null, &Query::Value) => Ok(()),
+      (&Schema::Null, &Query::Object(_)) => Err(Error::validation("Cannot deeply query null.", NO_PRIMITIVE_HINT)),
+      (&Schema::Boolean, &Query::Value) => Ok(()),
+      (&Schema::Boolean, &Query::Object(_)) => Err(Error::validation("Cannot deeply query a boolean.", NO_PRIMITIVE_HINT)),
+      (&Schema::Number{..}, &Query::Value) => Ok(()),
```

Comment:
I don't know either, but the tests pass 👌

Out of curiosity what smells to you?